Sources/NIOOpenSSL/SSLPKCS12Bundle.swift

```diff
@@ -0,0 +1,218 @@
+        certs.append(OpenSSLCertificate.fromUnsafePointer(pointer: actualCert))
+
+        for idx in 0..<certStackSize {
+            guard let stackCertPtr = CNIOOpenSSL_sk_X509_value(caCerts, idx) else {
+                preconditionFailure("Unable to get cert \(idx) from stack \(String(describing: caCerts))")
+            }
+            certs.append(OpenSSLCertificate.fromUnsafePointer(pointer: stackCertPtr))
+        }
+
+        self.certificateChain = certs
+        self.privateKey = OpenSSLPrivateKey.fromUnsafePointer(pointer: actualKey)
+
+    }
+
+    /// Create a `OpenSSLPKCS12Bundle` from the given bytes in memory,
+    /// optionally decrypting the bundle with the given passphrase.
+    ///
+    /// - parameters:
+    ///     - buffer: The bytes of the PKCS#12 bundle.
+    ///     - passphrase: The passphrase used for the bundle, as a sequence of UTF-8 bytes.
+    public init<Bytes: Collection>(buffer: [UInt8], passphrase: Bytes?) throws where Bytes.Element == UInt8 {
+        guard openSSLIsInitialized else { fatalError("Failed to initialize OpenSSL") }
+        
+        let p12 = buffer.withUnsafeBytes { pointer -> UnsafeMutablePointer<PKCS12>? in
+            let bio = BIO_new_mem_buf(UnsafeMutableRawPointer(mutating: pointer.baseAddress), CInt(pointer.count))!
+            defer {
+                BIO_free(bio)
+            }
+            return d2i_PKCS12_bio(bio, nil)
+        }
+        defer {
+            p12.map(PKCS12_free)
+        }
+
+        if let p12 = p12 {
+            try self.init(ref: p12, passphrase: passphrase)
+        } else {
+            throw OpenSSLError.unknownError(.init())
+        }
+    }
+
+    /// Create a `OpenSSLPKCS12Bundle` from the given bytes on disk,
+    /// optionally decrypting the bundle with the given passphrase.
+    ///
+    /// - parameters:
+    ///     - file: The path to the PKCS#12 bundle on disk.
+    ///     - passphrase: The passphrase used for the bundle, as a sequence of UTF-8 bytes.
+    public init<Bytes: Collection>(file: String, passphrase: Bytes?) throws where Bytes.Element == UInt8 {
+        guard openSSLIsInitialized else { fatalError("Failed to initialize OpenSSL") }
+
+        let fileObject = fopen(file, "rb")
+        defer {
+            fclose(fileObject)
+        }
+
+        let p12 = d2i_PKCS12_fp(fileObject, nil)
+        defer {
+            p12.map(PKCS12_free)
+        }
+
+        if let p12 = p12 {
+            try self.init(ref: p12, passphrase: passphrase)
+        } else {
+            throw OpenSSLError.unknownError(.init())
+        }
+    }
+
+    /// Create a `OpenSSLPKCS12Bundle` from the given bytes on disk,
+    /// assuming it has no passphrase.
+    ///
+    /// If the bundle does have a passphrase, call `init(file:passphrase:)` instead.
+    ///
+    /// - parameters:
+    ///     - file: The path to the PKCS#12 bundle on disk.
+    public init(file: String) throws {
+        try self.init(file: file, passphrase: nil as [UInt8]?)
```

Comment:
not sure but I prefer `Optional<[UInt8]>.none` or `[UInt8]?.none` instead of `nil as [UInt8]?` but feel free to ignore